readme.md
# LuckyNodeStudy

> LuckyNodeStudy是一个包含了很多有意思的学习node的开源项目。(LuckyNodeStudy is a project for studying node with many interesting example)

## browserify：nodejs平台代码运用在浏览器平台案例

> 通过本案例，您可以体会到browserify一种独特的思路：nodejs平台代码打包为web使用js的一种打包方案，除seajs/requirejs浏览器依赖外的独立js资源打包方案
项目中也采用了多种好用的功能能够
-编译less、更改后自动编译
-Browserify编译nodejs平台下js为独立js
-更改less、js、html自动刷新页面
-web服务器

### 安装nodejs
nodejs安装nodejs程序安装

### 安装gulp
运行 install-gulp.bat

### 安装node引用组建
运行 install.bat

### 运行
运行 run.bat

### 浏览器打开地址
- 浏览器输入：http://127.0.0.1:12580/browserify/html/page_index.html


## node基础篇

### 案例一：人生的意义[study readline]

> 通过使用nodejs的readline，来体会nodejs为开发者提供的命令行传入命令的界面，更有助于命令行程序的控制

> $ node node-study-whatIsYourLife



### 案例二：基于net的最简易web服务器[study net]

> 本案例是基于node最基础的net模块，手动输出http报头，相比于高大上的http等进阶模块，本案例更能够让开发者体会到最基础的tcp连接的真谛

> $ node node-study-net-webserver

> port:12570

### 案例三：基于socket.io的websocket服务器/客户端案例[study websocket server/client]

> 本案例是集成目前web端比较不错的js通信组件socket.io，实现websocket实时通信案例

> $ node node-study-WebSocketServer

> port:12571

### 案例四：异步编程：一步一个脚印[study async]

> node中很多基于异步编程：比如读取文件、网络请求等，很多时候需要一个一个按照顺序执行，这就像一步一个脚印，走好前一步才能走到下一步

> $ node node-study-async

> tags
- asysc.series
- asysc.waterfall

### 案例五：事件驱动，注册事件、响应事件。起床、（听新闻）、吃早饭、去上班[study events]

> LuckyNodeStudy之event.EventEmitter事件驱动：注册事件on，需要执行的时候执行事件emit。如：注册起床、听收音机 、上班事件需要做的事情，到时间按部就班执行。

> node中使用on注册事件，使用emit执行事件，现在很多代码语法都按照这个思路去做。比如socket.io也是语法上一样。
var events = require("events");
var event = new events.EventEmitter();
//reg event
event.on('xx', function() {
    //do something
}
//do something
event.emit("xx");
> $ node node-study-event-EventEmitter

## react+seajs+backbone+gulp环境

> 实现动态编译jsx，web服务器，实现 react+seajs+jquery+backbone+understore
> 插件：gulp-react




changelog.md
# LuckyNodeStudy

> LuckyNodeStudy是一个包含了很多有意思的学习node的开源项目。(LuckyNodeStudy is a project for study node with many interesting example for you)其中包括

## 2016.9.26
### 11:14
- 创建LuckyNodeStudy项目，增加案例一：人生的意义[study readline]，文件：node-study-whatIsYourLife.js

### 15:40
- 增加node-study-net-webserver.js简易web服务器示例
- 增加node-study-WebSocket基于socket.io的websocket演示demo

## 2016.9.28
- 加node异步编程案例，async的series、waterfall用法

## 2016.9.30
### browserify：nodejs平台代码运用在浏览器平台案例

> 通过本案例，您可以体会到browserify一种独特的思路：nodejs平台代码打包为web使用js的一种打包方案，除seajs/requirejs浏览器依赖外的独立js资源打包方案

### 功能模块
-编译less、更改后自动编译
-Browserify编译nodejs平台下js为独立js
-更改less、js、html自动刷新页面
-web服务器

### 安装nodejs
nodejs安装nodejs程序安装

### 安装gulp
运行 install-gulp.bat

### 安装node引用组建
运行 install.bat

### 运行
运行 run.bat

### 浏览器打开地址
- 浏览器输入：http://127.0.0.1:12580/browserify/html/page_index.html

## 2016.10.2
### 19:50
- 增加mac/linux下安装、编译脚本:install.sh/run.sh
- 修复package.json插件不全问题,原先运行比过年没有完全安装执行
- 如遇到无法安装需要sudo 管理员权限

## 2016.10.4
### 11:20
- 增加node的EventEmitter事件驱动用法案例：node-study-event-EventEmitter.js

## 2016.10.10
### 17:09
- react+seajs+backbone
>1. jquery 2. underscore 3. backbone 4. reactjs 5. seajs

- jsx动态监听编译为js
- 运行方式：$ gulp reactTask
- 浏览器打开 http://127.0.0.1:12550

Gulpfile.js
var loader = require('gulp-load-plugins')();
var gulp = require("gulp");
var less = require('gulp-less');
var clean = require('gulp-clean');
var browserify = require("gulp-browserify");
var path = require("path");
//更名
var rename = require('gulp-rename');
//压缩js
var uglify = require('gulp-uglify');
//实时检测
var connect = require('gulp-connect');
//node报错友好提示
var plumber = require('gulp-plumber');

var gulpReact = require('gulp-react');


var config = {
    browserify: "browserify",
    browserifyPort: 12580,
    react: "react",
    reactPort: 12550
};

//清理css文件夹
gulp.task('clean-browserify-css', function () {
    return gulp;
    return gulp.src(config.browserify + '/css/**/*.*', {
        read: false
    }).pipe(clean({force: true}));
});

gulp.task('browserify-less', ['clean-browserify-css'], function () {
    return gulp.src([config.browserify + '/less/**/*.less'])
        .pipe(less({
            project: [path.join(__dirname, 'less', 'includes')]
        }))
        .pipe(gulp.dest(config.browserify + '/css/'))
});
//browserify-test
gulp.task('browserify-js', function () {
    return gulp.src(config.browserify + '/script/pages/*.*')
        .pipe(plumber())
        .pipe(browserify())
        .pipe(gulp.dest(config.browserify + '/script/dist'))
        .pipe(rename({suffix: '.min'}))
        .pipe(uglify({
            mangle: {except: ['require', '$', 'define', 'module', "exports", "use", "config"]}
        }))
        .pipe(gulp.dest(config.browserify + '/script/dist'))
});

//创建服务器
gulp.task('browserify-connect', function () {
    connect.server({
        host: '127.0.0.1',           //Server host
        root: "./",
        port: config.browserifyPort,
        debug: true,
        livereload: true
    });
});

gulp.task('watcher-browserify-less', ['browserify-less'], function () {
    gulp.watch(config.browserify + '/less/**/*.less', ['browserify-less', "browserify-js-reload"]); //当less文件发生改变时，自动编译成css
});
//重新reload
gulp.task('browserify-less-reload', ['browserify-less'], function () {
    gulp.src([config.browserify + '/html/**/*.html'])
        .pipe(connect.reload());

});
gulp.task('watcher-browserify-js', ['browserify-js'], function () {
    gulp.watch([config.browserify + '/script/pages/**/*.js'], ['browserify-js', 'browserify-less-reload']);
});

//重新reload
gulp.task('browserify-js-reload', ['clean-browserify-css', 'browserify-js'], function () {
    gulp.src([config.browserify + '/html/**/*.html'])
        .pipe(connect.reload());
});

//重新reload
gulp.task('browserify-html-reload', function () {
    gulp.src([config.browserify + '/html/**/*.html'])
        .pipe(connect.reload());
});

//重新reload
gulp.task('watcher-browserify-html', function () {
    gulp.watch([config.browserify + '/html/**/*.html'], ['browserify-html-reload']);
});

//2016.10.10 react服务器搭建
//创建服务器
gulp.task('react-connect', function () {
    connect.server({
        host: '127.0.0.1',           //Server host
        root: "./",
        port: config.reactPort,
        debug: true,
        livereload: true
    });
});


gulp.task('default', ['watcher-browserify-js', 'watcher-browserify-less', 'watcher-browserify-html', 'browserify-connect'], function () {
    console.log("default task done");
});

//重新reload
gulp.task('watcher-react-jsx', function () {
    gulp.watch([config.react + '/myreact/**/*.*'], ['react-build']);
});

gulp.task('react-build', function () {
    return gulp.src([config.react + '/myreact/**/*.jsx'])
        .pipe(gulpReact())
        .pipe(gulp.dest(config.react + '/static/'));
});

gulp.task('reactTask', ['react-connect', 'react-build', 'watcher-react-jsx'], function () {
    console.log("react task done");
});
